File: ApiUtils.java
package com.apitest.utils;

import io.restassured.response.Response;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.testng.Assert;

/**
 * ApiUtils — assertion helpers and response-inspection shortcuts
 * that keep test methods clean and DRY.
 */
public class ApiUtils {

    private static final Logger log = LogManager.getLogger(ApiUtils.class);

    // ── Status code assertions ────────────────────────────────────────────────

    public static void assertStatusCode(Response response, int expected) {
        int actual = response.getStatusCode();
        log.info("Asserting status code — expected: {}, actual: {}", expected, actual);
        Assert.assertEquals(actual, expected,
                "Unexpected status code. Body: " + response.getBody().asString());
    }

    public static void assertOk(Response response)          { assertStatusCode(response, 200); }
    public static void assertCreated(Response response)     { assertStatusCode(response, 201); }
    public static void assertNoContent(Response response)   { assertStatusCode(response, 204); }
    public static void assertBadRequest(Response response)  { assertStatusCode(response, 400); }
    public static void assertUnauthorized(Response response){ assertStatusCode(response, 401); }
    public static void assertForbidden(Response response)   { assertStatusCode(response, 403); }
    public static void assertNotFound(Response response)    { assertStatusCode(response, 404); }

    // ── Response-time assertions ──────────────────────────────────────────────

    /**
     * Asserts the response was received within {@code maxMs} milliseconds.
     * Logs a warning rather than failing when the limit is only slightly breached.
     */
    public static void assertResponseTime(Response response, long maxMs) {
        long actual = response.getTime();
        log.info("Response time: {} ms (limit: {} ms)", actual, maxMs);
        if (actual > maxMs) {
            String msg = String.format("Response time %d ms exceeded limit of %d ms", actual, maxMs);
            Assert.fail(msg);
        }
    }

    public static void assertFastResponse(Response response) {
        assertResponseTime(response, 2_000);
    }

    // ── Field extraction helpers ──────────────────────────────────────────────

    public static <T> T extractField(Response response, String jsonPath, Class<T> type) {
        T value = response.jsonPath().getObject(jsonPath, type);
        log.info("Extracted '{}' = {}", jsonPath, value);
        Assert.assertNotNull(value, "Field '" + jsonPath + "' was null in response");
        return value;
    }

    public static String extractString(Response response, String jsonPath) {
        return extractField(response, jsonPath, String.class);
    }

    public static Integer extractInt(Response response, String jsonPath) {
        return extractField(response, jsonPath, Integer.class);
    }

    // ── Content-type assertion ────────────────────────────────────────────────

    public static void assertJsonContentType(Response response) {
        String ct = response.getContentType();
        log.info("Content-Type: {}", ct);
        Assert.assertTrue(ct.contains("application/json"),
                "Expected JSON content type, got: " + ct);
    }

    // ── Generic not-null body assertion ──────────────────────────────────────

    public static void assertBodyNotEmpty(Response response) {
        String body = response.getBody().asString();
        Assert.assertNotNull(body, "Response body is null");
        Assert.assertFalse(body.isEmpty(), "Response body is empty");
    }
}

File: src/utils/api_utils.py
"""
api_utils.py — assertion helpers and response-inspection shortcuts
that keep test functions clean and DRY. Mirrors the Java ApiUtils.java.
"""

import logging
from requests import Response

log = logging.getLogger("api_utils")


# ── Status code assertions ──────────────────────────────────────────────────

def assert_status_code(response: Response, expected: int):
    actual = response.status_code
    log.info(f"Asserting status code — expected: {expected}, actual: {actual}")
    assert actual == expected, (
        f"Unexpected status code. Expected {expected}, got {actual}. "
        f"Body: {response.text}"
    )


def assert_ok(response: Response):           assert_status_code(response, 200)
def assert_created(response: Response):      assert_status_code(response, 201)
def assert_no_content(response: Response):   assert_status_code(response, 204)
def assert_bad_request(response: Response):  assert_status_code(response, 400)
def assert_unauthorized(response: Response): assert_status_code(response, 401)
def assert_forbidden(response: Response):    assert_status_code(response, 403)
def assert_not_found(response: Response):    assert_status_code(response, 404)


# ── Response-time assertions ─────────────────────────────────────────────────

def assert_response_time(response: Response, max_ms: int):
    """Asserts the response was received within max_ms milliseconds."""
    actual_ms = response.elapsed.total_seconds() * 1000
    log.info(f"Response time: {actual_ms:.0f} ms (limit: {max_ms} ms)")
    assert actual_ms <= max_ms, (
        f"Response time {actual_ms:.0f} ms exceeded limit of {max_ms} ms"
    )


def assert_fast_response(response: Response):
    assert_response_time(response, 2_000)


# ── Field extraction helpers ────────────────────────────────────────────────

def extract_field(response: Response, key: str):
    value = response.json().get(key)
    log.info(f"Extracted '{key}' = {value}")
    assert value is not None, f"Field '{key}' was null in response"
    return value


def extract_string(response: Response, key: str) -> str:
    return str(extract_field(response, key))


def extract_int(response: Response, key: str) -> int:
    return int(extract_field(response, key))


# ── Content-type assertion ──────────────────────────────────────────────────

def assert_json_content_type(response: Response):
    ct = response.headers.get("Content-Type", "")
    log.info(f"Content-Type: {ct}")
    assert "application/json" in ct, f"Expected JSON content type, got: {ct}"


# ── Generic not-null body assertion ─────────────────────────────────────────

def assert_body_not_empty(response: Response):
    body = response.text
    assert body is not None, "Response body is null"
    assert body != "", "Response body is empty"
